README.md
[![prokeys logo](https://i.stack.imgur.com/HrCnC.png)](https://chrome.google.com/webstore/detail/prokeys/ekfnbpgmmeahnnlpjibofkobpdkifapn)

**Developed by**: Aquila Softworks ([Gaurang Tandon](https://github.com/GaurangTandon) and [Varun Singh](https://github.com/iWrote))  
**Latest version**: 3.4.3
**Status**: active, stable build (last update 15/Sep/2018 IST)

ProKeys is a completely free, [Google Chrome](https://chrome.google.com/webstore/detail/prokeys/ekfnbpgmmeahnnlpjibofkobpdkifapn) and [Opera](https://addons.opera.com/en/extensions/details/prokeys/?display=en) extension that lets you be productive in online text-related work with its number of features like:

[More info on all these in Help section inside the app]

1.  **Snippets** - define custom abbreviations, and retrieve the text associated with them simply by pressing the hotkey (default: Shift+Space). Example: "brb" can expand to "Be right back!"
2.  **Placeholders** - are fields in the snippet body that can be given dynamic values on using the snippet.
3.  **Mathomania** - do math without leaving your text editor, and without calculators. Example: "[[ 15% * 600 =]]" gives "90" and "[[ (5+6) * 15^2 =]]" gives "2475".
4.  **Clipboard access** - insert `[[%p]]` anywhere in your snippet body. It will automatically be substituted with the current clipboard data.
5.  **Snippet embedding**: you can embed one or more snippets inside another snippet, simply by doing `[[%s(snip_name)]]`.
6.  **Auto-Insert** - Quotes ('"') and braces ('(', '[', '{') are auto-completed, and you can specify your custom insert-it-for-me also (like, inserting a '>' on typing of a '<') in the settings.
7.  **Date/Time Macros** - embed short symbols inside the parentheses of '[[%d()]]' inside a snippet body and it will auto replace the symbol with the current date and time related value. A sample snippet as well as a guide has been provided (in Help section inside the app). **Date/Time Arithmetic** - use the `+` or `-` sign to move forward or backward in time in date/time macros. Supports both relative as well as independent evaluation.
8.  **Omnibox support** - search through your snippets and use them right in the browser address bar!
9.  **URL Macro** - insert customizable parts of the current tab URL into the snippet.
10. **Variables** - built-in variables holding dynamic values for "date", "time" and (browser) "version"
11. **Tab Key** - The tab key can be made to insert 4 spaces, instead of it's default function, thus speeding up work.
12. **Context menu access** - for blocking sites and inserting snippets.
13. **Sync/Local storage** - Use whichever you prefer.

## Libraries/Code sources

Except for the feather-weight open-source QuillJS rich text editor, I only utilized vanilla JS for super performance and less app size. Also Eric Meyer's CSS reset stylesheet - public domain - has been used with modifications.

Some icon images have been sourced from Font Awesome 5. The license is at [this link](https://fontawesome.com/license/free). No changes (except scaling) were made to these images.

## Technical details

Are shared on [here](http://electricweb.org/chrome-extension-tutorial-snippets) in a long, three post series. In brief: this extension uses content scripts which are injected into every page and (what follows is for the snippet functionality only) every time the user presses the hotkey, a check is made for the preceding text, which if matches the name of any of the snippets, is substituted with the snippet body. A subsequent test for placeholders is made as well.

## Contribute

First of all, thanks for your contribution! Every small bit of it counts! You can:

1.  [Create a new issue](https://github.com/GaurangTandon/ProKeys/issues/new) for bugs, feature requests, and enhancements.
2.  [Write a review](https://chrome.google.com/webstore/detail/prokeys/ekfnbpgmmeahnnlpjibofkobpdkifapn/reviews) (or [anonymous feedback](https://docs.google.com/forms/d/1DcwQB5vnNCH0pP_Y-wVvOF6gsI0gaXGPPngctb4tCdA/viewform?usp=send_form)), or [get support](https://chrome.google.com/webstore/detail/prokeys/ekfnbpgmmeahnnlpjibofkobpdkifapn/support) with some technical problems.
3.  Fork the repo, make changes, and submit a pull request, describing the changes made.
4.  Help me translate ProKeys to your native language.
5.  Share the word about ProKeys with people!
6.  Monetary donations are gratefully accepted at email 1gaurangtandon@gmail.com using [PayPal](https://www.paypal.com/myaccount/transfer/buy)

## Development

To setup this repo, clone it, `cd` into it, and then run `npm run build`. This will create a unminified development build in `./dist` which you can then load into Chrome. This step is required because Chrome does not yet natively support `import`/`export` syntax for Chrome extensions.

We use both global variables and `import`/`export` ones. The `import`/`export` ones are either functions or unmodified constants. Only those variables are declared under `window` or `pk` which need to be modified in different scripts.

**Contact us** - prokeys.feedback@gmail.com - to discuss anything related to the above if you want to.

## [Known Issues](https://docs.google.com/document/d/1_MHKm1jtpJCWgksfbUdufExRFlF81S-IuTz1Czu7gOI/edit?usp=sharing)

## [Change Log](https://github.com/GaurangTandon/ProKeys/blob/master/change_log.md)


change_log.md
## Change log

**This Change log has been deprecated since version 3.4.\*. Please head over to [Releases](https://github.com/GaurangTandon/ProKeys/releases) to read the latest changes.**

Helps organize exactly what things happened in which version.

---

**3.3.1** - 25/Jul/2018

-   embed snippets support
-   more robust data restore functionality (ignore unknown props/set to default known props if absent)
-   fix regressions of time function not working and of auto-insert mis-wrapping in whitespace
-   fix regression snippet insertion dead through both hotkey and ctx

**3.2.1** - 21/Jul/2018

-   regression of data restore missing property

**3.2.0** - 20/Jul/2018

-   Snippet expansion inside the address bar (omnibox) is now supported.
-   Added a new snippet macro to insert the current browser URL (or parts of it).
-   Added a clone button that lets you duplicate a snippet or a folder.
-   Pressing an auto-insert with some text selected now wraps the selection inside of it.
-   Using auto-inserts no longer messes up text wrap/formatting in Gmail (and other rich text editors).
-   Auto-inserts are no longer inserted for specific patterns like :-(

**3.1.3** - 17/Mar/2017

-   NodeList prorotype modification now works across iframes - by attaching props to them individually (see [SO](http://stackoverflow.com/questions/42825990/extending-prototype-of-dom-elements-inside-iframes) question)
-   Hence the bug for Evernote, Salesforce, Lithium, and all other iframe elements is fixed

**3.1.2** - 15/Mar/2017

-   Links added by user without any protocol get "http://" prepended by default (as Chrome expects all links to have protocols).
-   Bold and italics now retain tag used by user.
-   Links are now properly displayed in textarea.
-   Nice formatted display of lists, blockquotes, and `<pre>` blocks in snippet list.
-   Further significant improvement to rich text usage (#153)

**3.1.1** - 06/Mar/2017

-   Added FAQ section and Sync troubleshooting section in Help page
-   Added survey form for uninstalls
-   Added significant support for rich text usage (YAY!)
-   Console.log error in webpages after update.
-   made export/backup popup concise
-   added `debounce` to search field, window resize handler
-   searching ignores tags that might overlap with matching text
-   other minor issues

**3.1.0** - 13/Feb/2017

-   Added completely new rich text editor (YAY!) in snippets page.
-   Compressed all images to save 50KB!
-   Bug fixed: storage bytes (under Snippets heading) were shown as "906.0B" instead of "906B"
-   designed better the Help page (cleaner font, bullets, more colors)
-   redesigned blocked sites display
-   fixed auto-insert display code
-   better handling duplicate snippets and folders in restore data (ASSUMPTION: the input data for restore has no duplicates in itself) by offering, in case of duplicate data, to keep either both imported and existing snippets or one of them. Also, can merge duplicate folder's contents.
-   expansion/contraction of folders in `.selectList`
-   allow option to match snippet by whole word in Settings (separated by delimiter - custom list user editable)
-   removed the option to choose type of data ("entire data" vs "only snippets") by enabling manual detection for the type of data (entire data is an object and snippets is an array)
-   Selectors in site block modal css were interfering with normal webpages.
-   Context menu snippet insertion was non-functional.
-   No feature was working in Gmail subject field
-   "Choose file" link in Restore popup was non-functional.

**3.0.0.1** - 27/May/2016

-   linted the js files
-   fixed two embarrrasing typos in change log
-   fixed the error in Snip.fromObject where snip.timestamp was undefined

**3.0.0** - 27/May/2016

-   made snippet handling OOP, added folder suppet
-   logo redesign (thanks [Varun](https://github.com/iWrote)!)
-   removed snippets popup, listing snippets in options page now
-   completely redesigned the options page, made responsive till 200% zoom
-   fixed: https://github.com/GaurangTandon/ProKeys/issues/3
-   first install should open popup box directly and give a greeting message/explaining 3.0.0 changes - done

js/snippetWork.js
/* global Data */
// eslint-disable-next-line no-unused-vars
/* global $containerSnippets, $panelSnippets */

import {
    q, qCls, qClsSingle, qId, Q, debounce, SHOW_CLASS,
} from "./pre";
import {
    Folder, Snip, Generic, DualTextbox,
} from "./snippetClasses";
import { saveSnippetData } from "./commonDataHandlers";

let validateSnippetData,
    validateFolderData,
    toggleSnippetEditPanel,
    toggleFolderEditPanel,
    dualSnippetEditorObj;
/**
 * @param {String} type generic snip or folder
 * @returns {Function} the handler for given type
 */
function handlerSaveObject(type) {
    const validationFunc = type === Generic.SNIP_TYPE ? validateSnippetData : validateFolderData;
    // once checked, now can mutate type as per need
    type = type[0].toUpperCase() + type.substr(1).toLowerCase();

    return function () {
        return validationFunc((oldName, name, body, newParentfolder) => {
            if (oldName) {
                const oldObject = Data.snippets[`getUnique${type}`](oldName),
                    oldParentFolder = oldObject.getParentFolder();

                if (newParentfolder.name !== oldParentFolder.name) {
                    // first move that object; before changing its name/body
                    const movedObject = oldObject.moveTo(newParentfolder);

                    movedObject.name = name;
                    if (type === "Snip") {
                        movedObject.body = body;
                    }

                    window.latestRevisionLabel = `moved "${name}" ${movedObject.type} to "${
                        newParentfolder.name
                    }"`;
                    saveSnippetData(undefined, newParentfolder.name, name);
                } else {
                    oldParentFolder[`edit${type}`](oldName, name, body);
                    saveSnippetData(undefined, newParentfolder.name, name);
                }
            } else {
                newParentfolder[`add${type}`](name, body);
                saveSnippetData(undefined, newParentfolder.name, name);
            }
        });
    };
}

/**
 * setup the folder or snippet edit panel
 * attaches handlers to div's
 * @param {String} type generic snip or folder
 * @returns {Function} triggered for display of edit panel
 */
function setupEditPanel(type) {
    const $panel = qClsSingle(`panel_${type}_edit`),
        saveHandler = handlerSaveObject(type),
        nameElm = $panel.q(".name input");
    let $parentFolderDIV;

    $panel.on("keydown", (event) => {
        // ctrl/cmd-enter
        if (event.keyCode === 13 && (event.ctrlKey || event.metaKey)) {
            saveHandler();
        }

        // escape
        if (event.keyCode === 27) {
            $panel.qClsSingle("close_btn").click();
        }
    });

    nameElm.on("keydown", (event) => {
        // the ! ensures that both the $panel and the nameElm handlers don't fire simultaneously
        if (event.keyCode === 13 && !(event.ctrlKey || event.metaKey)) {
            saveHandler();
        }
    });

    function highlightInFolderList(folderElm, name) {
        const $folderNames = folderElm.Q("p");
        let folderUsable = null;

        for (const folder of $folderNames) {
            if (folder.html() === name) {
                folderUsable = folder;
                folder.addClass("selected");
                break;
            }
        }

        if (folderUsable) {
            $parentFolderDIV = folderUsable.parent("div");
            while (!$parentFolderDIV.classList.contains("selectList")) {
                $parentFolderDIV.removeClass("collapsed");
                $parentFolderDIV = $parentFolderDIV.parent("div");
            }
        }
    }

    return function (object, isSavingSnippet) {
        const headerSpan = $panel.q(".header span"),
            folderElm = $panel.q(".folderSelect .selectList"),
            folderPathElm = $panelSnippets.q(".folder_path :nth-last-child(2)"),
            // boolean to tell if call is to edit existing snippet/folder
            // or create new snippet
            isEditing = !!object,
            isSnip = type === "snip",
            errorElements = qCls("error");

        $panelSnippets.toggleClass(SHOW_CLASS);
        $panel.toggleClass(SHOW_CLASS);
        if (isEditing) {
            $panel.removeClass("creating-new");
        } else {
            $panel.addClass("creating-new");
        }

        /* cleanup  (otherwise abnormal rte swap window.alerts are received from
                userAllowsToLoseFormattingOnSwapToTextarea method when we
                later would load another snippet) */
        dualSnippetEditorObj.setPlainText("").setRichText("");

        // had clicked the tick btn
        if (isSavingSnippet) {
            return;
        }

        // reset
        folderElm.html("");

        // cannot nest a folder under itself
        folderElm.appendChild(
            isEditing && type === "folder"
                ? Data.snippets.getFolderSelectList(object.name)
                : Data.snippets.getFolderSelectList(),
        );

        if (isEditing) {
            const parent = object.getParentFolder();
            highlightInFolderList(folderElm, parent.name);
        } else {
            highlightInFolderList(folderElm, folderPathElm.html());
        }

        headerSpan.html((isEditing ? "Edit " : "Create new ") + type);

        if (errorElements) {
            errorElements.removeClass(SHOW_CLASS);
        }

        // defaults
        if (!isEditing) {
            object = {
                name: "",
                body: "",
            };
        }

        nameElm.text(object.name).focus();
        nameElm.dataset.name = object.name;
        if (isSnip) {
            dualSnippetEditorObj.switchToDefaultView(object.body).setShownText(object.body);
        }
    };
}

/** functions common to snip and folder
 * @param {String} panelName snip or folder only
 * @returns {Function} extracts text from elms, passes them for validation, callsback
 */
function commonValidation(panelName) {
    const panel = qClsSingle(`panel_${panelName}_edit`);

    function manipulateElmForValidation(elm, validateFunc, errorElm) {
        const text = elm ? elm.value : dualSnippetEditorObj.getShownTextForSaving(),
            textVld = validateFunc(text),
            textErrorElm = errorElm || elm.nextElementSibling;
        let isTextValid = textVld === "true";

        // when we are editing snippet/folder it does
        // not matter if the name remains same
        if (
            textVld === Generic.getDuplicateObjectsText(text, panelName)
            && elm.dataset.name === text
        ) {
            isTextValid = true;
        }

        if (isTextValid) {
            textErrorElm.removeClass(SHOW_CLASS);
        } else {
            textErrorElm.addClass(SHOW_CLASS).html(textVld);
        }

        return [text, isTextValid];
    }

    return function (callback) {
        const nameElm = panel.q(".name input"),
            selectList = panel.qClsSingle("selectList"),
            folder = Folder.getSelectedFolderInSelectList(selectList),
            isSnippet = /snip/.test(panel.className);
        let name,
            body;

        if (isSnippet) {
            name = manipulateElmForValidation(nameElm, Snip.isValidName);
            body = manipulateElmForValidation(undefined, Snip.isValidBody, panel.q(".body .error"));
        } else {
            name = manipulateElmForValidation(nameElm, Folder.isValidName);
        }

        const allValid = name[1] && (isSnippet ? body[1] : true),
            oldName = nameElm.dataset.name;

        if (allValid) {
            if (isSnippet) {
                toggleSnippetEditPanel(undefined, true);
            } else {
                toggleFolderEditPanel(undefined, true);
            }

            callback(oldName, name[0], body && body[0], folder);
        }
    };
}

export function initSnippetWork() {
    // define element variables
    const $searchBtn = qClsSingle("search_btn"),
        $searchPanel = qClsSingle("panel_search"),
        $searchField = q(".panel_search input[type=text]"),
        $closeBtn = qCls("close_btn"),
        $snippetSaveBtn = q(".panel_snip_edit .tick_btn"),
        $folderSaveBtn = q(".panel_folder_edit .tick_btn"),
        $addNewBtns = Q(".panel_snippets [class^=\"add_new\"]"),
        $sortBtn = q(".panel_snippets .sort_btn"),
        $sortPanel = qClsSingle("panel_sort"),
        // the button that actually initiates sorting
        $sortPanelBtn = q(".panel_sort input[type=button]"),
        $bulkActionBtn = q(".panel_snippets .checkbox_btn"),
        $bulkActionPanel = q(".panel_snippets .panel_bulk_action"),
        folderPath = qClsSingle("folder_path"),
        $selectList = qCls("selectList");

    validateSnippetData = commonValidation(Generic.SNIP_TYPE);
    validateFolderData = commonValidation(Generic.FOLDER_TYPE);
    toggleSnippetEditPanel = setupEditPanel(Generic.SNIP_TYPE);
    toggleFolderEditPanel = setupEditPanel(Generic.FOLDER_TYPE);
    dualSnippetEditorObj = new DualTextbox(qId("body-editor"));

    /**
     * Delegated handler for edit, delete, clone buttons
     */
    $containerSnippets.on("click", (e) => {
        const node = e.target,
            container = node.parentNode,
            objectElm = container.parentNode;
        let obj;

        if (!objectElm || !/buttons/.test(container.className)) {
            return true;
        }

        if (node.matches("#snippets .panel_content .edit_btn")) {
            obj = Generic.getObjectThroughDOMListElm(objectElm);
            if (Folder.isFolder(obj)) {
                toggleFolderEditPanel(obj);
            } else {
                toggleSnippetEditPanel(obj);
            }
        } else if (node.matches("#snippets .panel_content .delete_btn")) {
            deleteOnClick.call(objectElm);
        } else if (node.matches("#snippets .panel_content .clone_btn")) {
            cloneBtnOnClick.call(objectElm);
        }

        return false;
    });

    folderPath.on("click", (e) => {
        const node = e.target;
        let folderName,
            folder;

        if (node.matches(".chevron")) {
            folder = Folder.getListedFolder();
            folder.getParentFolder().listSnippets();
        } else if (node.matches(".path_part")) {
            folderName = node.innerHTML;
            folder = Data.snippets.getUniqueFolder(folderName);
            folder.listSnippets();
        }
    });

    /**
     * @param {Element} $panel the snippet or folder panel being edited
     * @param {Element} $objectName the input element containing the object name
     * @returns {Boolean} if the text typed in the panel is edited (true)
     */
    function userHasEditedTextPresentInPanel($panel, $objectName) {
        const $body = $panel.qClsSingle("body"),
            objectName = $objectName.value,
            nameChanged = $objectName.dataset.name !== objectName;

        if (!$body || nameChanged) {
            return nameChanged;
        }

        if (objectName === "") {
            return false;
        }

        const body = dualSnippetEditorObj.getShownTextForSaving(),
            snip = Data.snippets.getUniqueSnip(objectName),
            bodyChanged = snip ? snip.body !== body : false;

        return bodyChanged;
    }

    function closePanel($panel) {
        $panel.removeClass(SHOW_CLASS);
        $panelSnippets.addClass(SHOW_CLASS);
    }

    $closeBtn.on("click", function () {
        const $panel = this.parent(".panel"),
            $objectName = $panel.q(".name input");

        if ($objectName && userHasEditedTextPresentInPanel($panel, $objectName)) {
            if (window.confirm("You have unsaved edits. Are you sure you wish to leave?")) {
                closePanel($panel);
            }
        } else {
            // not an edit panel, but rather some export popup
            closePanel($panel);
        }
    });

    $snippetSaveBtn.on("click", handlerSaveObject(Generic.SNIP_TYPE));
    $folderSaveBtn.on("click", handlerSaveObject(Generic.FOLDER_TYPE));

    function deleteOnClick() {
        const object = Generic.getObjectThroughDOMListElm(this),
            { name, type } = object,
            isSnip = type === Generic.SNIP_TYPE,
            warning = isSnip ? "" : " (and ALL its contents)";
        let folder;

        if (window.confirm(`Delete '${name}' ${type}${warning}?`)) {
            folder = object.getParentFolder();

            object.remove();
            this.parentNode.removeChild(this);

            window.latestRevisionLabel = `deleted ${type} "${name}"`;

            saveSnippetData(undefined, folder.name);
        }
    }

    function cloneBtnOnClick() {
        const clickedObject = Generic.getObjectThroughDOMListElm(this),
            objectClone = clickedObject.clone();
        window.latestRevisionLabel = `cloned ${clickedObject.type} "${clickedObject.name}"`;

        // keep the same snippet highlighted as well (object.name)
        // so that user can press clone button repeatedly
        saveSnippetData(undefined, objectClone.getParentFolder().name, [
            clickedObject.name,
            objectClone.name,
        ]);
    }

    $selectList.on("click", function (e) {
        const clickedNode = e.target,
            classSel = "selected",
            collapsedClass = "collapsed";
        let otherNodes,
            $containerDIV;

        if (clickedNode.tagName === "P") {
            // do not use $selectList
            // as it is a NodeList
            $containerDIV = clickedNode.parentNode;
            $containerDIV.toggleClass(collapsedClass);

            otherNodes = this.qCls(classSel);
            if (otherNodes) {
                otherNodes.removeClass(classSel);
            }
            clickedNode.addClass(classSel);
        }
    });

    // for searchBtn and $searchPanel
    // $addNewBtn and $addNewPanel
    // and other combos
    function toggleBtnAndPanel(btn, panel) {
        const existingPanel = q(".sub_panel.show");
        panel.addClass(SHOW_CLASS);
        if (existingPanel) {
            existingPanel.removeClass(SHOW_CLASS);
        }

        const existingBtn = q(".panel_btn.active");
        btn.addClass("active");
        if (existingBtn) {
            existingBtn.removeClass("active");
        }

        // we need these checks as another might have been clicked
        // to remove the search/checkbox panel and so we need to remove
        // their type of list
        if (!$searchPanel.hasClass(SHOW_CLASS) && Folder.getListedFolder().isSearchResultFolder) {
            Data.snippets.listSnippets();
        }

        // if checkbox style list is still shown
        if (
            $containerSnippets.q("input[type=\"checkbox\"]")
            && !$bulkActionPanel.hasClass(SHOW_CLASS)
        ) {
            Data.snippets
                .getUniqueFolder($bulkActionPanel.dataset.originalShownFolderName)
                .listSnippets();
        }
    }

    $sortBtn.on("click", () => {
        toggleBtnAndPanel($sortBtn, $sortPanel);
    });

    $sortPanelBtn.on("click", () => {
        let sortDir = $sortPanel.q(".sort-dir :checked").parentNode.innerText,
            sortType = $sortPanel.q(".sort-type :checked").parentNode.innerText;
        const descendingFlag = (sortDir = sortDir === "Descending"),
            lastFolderDOM = folderPath.lastChild.previousSibling,
            folder = Data.snippets.getUniqueFolder(lastFolderDOM.html());

        sortType = sortType === "Name" ? "alphabetic" : "date";

        folder.sort(sortType, descendingFlag);
        window.latestRevisionLabel = `sorted folder "${folder.name}"`;
        saveSnippetData(undefined, folder.name);
    });

    $addNewBtns.on("click", function () {
        if (/snip/i.test(this.className)) {
            toggleSnippetEditPanel();
        } else {
            toggleFolderEditPanel();
        }
    });

    $searchBtn.on("click", () => {
        toggleBtnAndPanel($searchBtn, $searchPanel);
        $searchField.html("").focus();
        // now hidden search panel, so re-list the snippets
        if (!$searchPanel.hasClass(SHOW_CLASS)) {
            Folder.getListedFolder().listSnippets();
        }
    });

    $searchBtn.attr("title", "Search for folders or snips");
    $searchField.on(
        "keyup",
        debounce(function searchFieldHandler() {
            const searchText = this.value,
                listedFolder = Folder.getListedFolder(),
                searchResult = listedFolder.searchSnippets(searchText);

            searchResult.listSnippets();
        }, 150),
    );

    (function bulkActionsWork() {
        let selectedObjects,
            DOMcontainer;
        const moveToBtn = $bulkActionPanel.q(".bulk_actions input:first-child"),
            deleteBtn = $bulkActionPanel.q(".bulk_actions input:last-child"),
            toggleAllButton = $bulkActionPanel.q(".selection_count input"),
            folderSelect = $bulkActionPanel.qClsSingle("folderSelect"),
            selectList = $bulkActionPanel.qClsSingle("selectList");

        function updateSelectionCount() {
            selectedObjects = DOMcontainer.Q("input:checked") || [];

            selectedObjects = selectedObjects.map((e) => {
                const div = e.nextElementSibling.nextElementSibling,
                    name = div.html(),
                    img = e.nextElementSibling,
                    type = img.src.match(/\w+(?=\.svg)/)[0];

                return Data.snippets.getUniqueObject(name, type);
            });

            $bulkActionPanel.q(".selection_count span").html(selectedObjects.length);

            $bulkActionPanel.Q(".bulk_actions input").forEach((elm) => {
                elm.disabled = !selectedObjects.length;
            });
        }

        $bulkActionBtn.on("click", function () {
            let originalShownFolderName,
                originalShownFolder;

            toggleBtnAndPanel(this, $bulkActionPanel);

            if ($bulkActionPanel.hasClass(SHOW_CLASS)) {
                originalShownFolderName = Folder.getListedFolderName();
                originalShownFolder = Data.snippets.getUniqueFolder(originalShownFolderName);
                DOMcontainer = Folder.insertBulkActionDOM(originalShownFolder);

                $bulkActionPanel.dataset.originalShownFolderName = originalShownFolderName;

                DOMcontainer.on("click", (event) => {
                    const nodeClicked = event.target,
                        parentGeneric = nodeClicked.matches(".generic")
                            ? nodeClicked
                            : nodeClicked.parent(".generic"),
                        inputCheckbox = parentGeneric.children[0];

                    if (nodeClicked.tagName !== "INPUT") {
                        inputCheckbox.checked = !inputCheckbox.checked;
                    }

                    updateSelectionCount();
                });

                updateSelectionCount();
                folderSelect.removeClass(SHOW_CLASS);
            }
        });

        toggleAllButton.on("click", () => {
            const checkboxes = DOMcontainer.Q("input");
            let allCheckedBoxesChecked = true,
                finalCheckState;

            checkboxes.some((checkbox) => {
                if (!checkbox.checked) {
                    allCheckedBoxesChecked = false;
                    return true;
                }
                return false;
            });

            finalCheckState = !allCheckedBoxesChecked;

            checkboxes.forEach((checkbox) => {
                checkbox.checked = finalCheckState;
            });

            updateSelectionCount();
        });

        // move to folder button
        moveToBtn.on("click", () => {
            let selectFolderName,
                selectedFolder,
                atleastOneElementMoved = false;

            if (!folderSelect.hasClass(SHOW_CLASS)) {
                Folder.refreshSelectList(selectList);
                folderSelect.addClass(SHOW_CLASS);
            } else {
                selectedFolder = Folder.getSelectedFolderInSelectList(selectList);
                selectFolderName = selectedFolder.name;
                selectedObjects.forEach((selObj) => {
                    if (selObj.canNestUnder(selectedFolder)) {
                        atleastOneElementMoved = true;
                        selObj.moveTo(selectedFolder);
                    } else {
                        window.alert(
                            `Cannot move ${selObj.type} "${selObj.name}" to "${
                                selectedFolder.name
                            }"`
                                + "; as it is the same as (or a parent folder of) the destination folder",
                        );
                    }
                });

                // do not list new folder if nothing was moved
                if (!atleastOneElementMoved) {
                    return;
                }

                window.latestRevisionLabel = `moved ${
                    selectedObjects.length
                } objects to folder "${selectFolderName}"`;

                saveSnippetData(
                    () => {
                        // hide the bulk action panel
                        $bulkActionBtn.click();
                    },
                    selectFolderName,
                    selectedObjects.map(e => e.name),
                );
            }
        });

        deleteBtn.on("click", () => {
            if (
                window.confirm(
                    `Are you sure you want to delete these ${selectedObjects.length} items? `
                        + "Remember that deleting a folder will also delete ALL its contents.",
                )
            ) {
                selectedObjects.forEach((selObj) => {
                    selObj.remove();
                });

                window.latestRevisionLabel = `deleted ${selectedObjects.length} objects`;

                saveSnippetData(() => {
                    // hide the bulk action panel
                    $bulkActionBtn.click();
                }, Folder.getListedFolderName());
            }
        });
    }());

    (function checkIfFirstTimeUser() {
        const $changeLog = qClsSingle("change-log"),
            $button = $changeLog.q("button"),
            // ls set by background page
            isUpdate = localStorage.extensionUpdated === "true";

        if (isUpdate) {
            $changeLog.addClass(SHOW_CLASS);

            $button.on("click", () => {
                $changeLog.removeClass(SHOW_CLASS);
                localStorage.extensionUpdated = "false";
            });
        }
    }());

    Data.snippets.listSnippets();
}
